[FLINK-36093][transform] fix preTransformoperator wrongly filters columns when multiple transform

MOBIN-F opened the pull request:
Currently, such transform rule could not work:

```yaml
transform:
projection: 'A' as result
filter: tag >= 0
projection: score as result
filter: tag < 0
```
Here, `score` column will be filtered out in PreTransform stage, since it wasn't referenced in the first transform rule. As a result, the following transform rule will fail since `score` does not exist in PostTransform operator.

leonardBang reviewed (approved):
Thanks @MOBIN-F for the contribution, I think current unit test is okay for me. Sure that future e2e test is welcome

leonardBang commented:
@MOBIN-F Could you help rebase this PR?

MOBIN-F commented:
> @MOBIN-F Could you help rebase this PR?
@leonardBang done~

MOBIN-F commented:
CI failure,It seems that this PR(https://github.com/apache/flink-cdc/pull/3696) should be merged first

leonardBang commented:
@MOBIN-F Could you kindly rebase again?

MOBIN-F commented:
> @MOBIN-F Could you kindly rebase again?

done~

MOBIN-F commented:
CI failure,I will recheck the code

MOBIN-F commented:
I previously ignored the scenario where projection was missing. I fixed it and added  integrated / E2e tests , Can you review it again? @yuxiqian 
In my CI Aaction,the tests passed
![image](https://github.com/user-attachments/assets/7dc9b443-4384-4d57-ba2c-f643eba0cee6)

yuxiqian commented on flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/PreTransformOperator.java:
```diff
@@ -342,27 +342,77 @@ private CreateTableEvent transformCreateTableEvent(CreateTableEvent createTableE
+            if (!transform.getSelectors().isMatch(tableId)) {
+                continue;
+            }
+            if (!transform.getProjection().isPresent()) {
+                processProjectionTransform(tableId, tableSchema, referencedColumnsSet, null);
+                hasMatchTransform = true;
+            } else {
```
Is it correct if we omit this branch and let it fallback to `if(!hasMatchTransform)` condition? Seems the handling logic is the same.

yuxiqian commented on flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/PreTransformOperator.java:
```diff
@@ -342,27 +342,77 @@ private CreateTableEvent transformCreateTableEvent(CreateTableEvent createTableE
+                                tableSchema.getColumns());
+                // update referenced columns of other projections of the same tableId, if any
+                referencedColumnsSet.addAll(referencedColumns);
+                hasAsteriskMap.put(tableId, false);
```
Should we check if `hasAsteriskMap[tableId]` is already `true`, perhaps set by a previous transform rule?

Add: I think it should not be a problem, since if someone puts asteriskful and non-asteriskful rules together, it will fail when column count changes from upstream anyway.

yuxiqian reviewed (commented):
Thanks for cleaning up `PreTransformOperator`! Looks much neater now, just left some trivial comments.

Also +@aiwenmo who is familiar with transform modules.

yuxiqian commented on flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/PreTransformOperator.java:
```diff
@@ -386,37 +436,16 @@ private Schema transformSchemaMetaData(
-                    return processProjection(transformProjection, transformFilter, dataChangeEvent);
-                }
-            }
+    private DataChangeEvent processDataChangeEvent(DataChangeEvent dataChangeEvent) {
+        if (!transforms.isEmpty()) {
+            return processProjection(dataChangeEvent);
         }
         return dataChangeEvent;
     }
```
Seems the logic is too simple and not worth being an independent method.

yuxiqian commented on flink-cdc-runtime/src/test/java/org/apache/flink/cdc/runtime/operators/transform/PreTransformOperatorTest.java:
```diff
@@ -526,4 +526,39 @@ void testMetadataTransform() throws Exception {
+    void testMultiTransform() throws Exception {
+        PreTransformOperator transform =
+                PreTransformOperator.newBuilder()
+                        .addTransform(
+                                CUSTOMERS_TABLEID.identifier(),
+                                "col1, upper(col1) col12",
+                                "col1 = 'col1'",
+                                "col2",
+                                "col12",
+                                "key1=value1,key2=value2")
+                        .addTransform(
+                                CUSTOMERS_TABLEID.identifier(),
+                                "col2, upper(col2) col12",
+                                "col1 != 'col1'",
+                                "col2",
+                                "col12",
+                                "key1=value1,key2=value2")
+                        .build();
```
Considering this bug is sequential specific, swapping rules to verify it would be nice.

yuxiqian commented on flink-cdc-runtime/src/test/java/org/apache/flink/cdc/runtime/operators/transform/PreTransformOperatorTest.java:
```diff
@@ -526,4 +526,39 @@ void testMetadataTransform() throws Exception {
+    @Test
+    void testMultiTransform() throws Exception {
+        PreTransformOperator transform =
+                PreTransformOperator.newBuilder()
```
Maybe we can also cover cases when projection is `*` or omitted here

MOBIN-F replied:
done~, also above
added some tests

aiwenmo commented:
> Thanks for cleaning up `PreTransformOperator`! Looks much neater now, just left some trivial comments.
> 
> Also +@aiwenmo who is familiar with transform modules.

Hi. @MOBIN-F. Thanks for the contribution. And I will assist in reviewing the code tonight~

MOBIN-F commented:
oceanbase tests failure,but it seems to have nothing to do with this PR

leonardBang commented:
@MOBIN-F Would you like to open a backport PR for `release-3.2` branch?

MOBIN-F commented:
> @MOBIN-F Would you like to open a backport PR for `release-3.2` branch?

OK
